editor/src/clj/editor/pose.clj

```diff
@@ -0,0 +1,255 @@
+(ns editor.pose
+  (:require [clojure.spec.alpha :as s]
+            [editor.math :as math])
+  (:import [javax.vecmath Matrix4d]))
+
+(set! *warn-on-reflection* true)
+(set! *unchecked-math* :warn-on-boxed)
+
+(s/def ::num number?)
+(s/def ::num3 (s/tuple ::num ::num ::num))
+(s/def ::num4 (s/tuple ::num ::num ::num ::num))
+(s/def ::translation ::num3)
+(s/def ::rotation ::num4)
+(s/def ::scale ::num3)
+(s/def ::euler-rotation ::num3)
+
+(defrecord Pose [translation rotation scale])
+
+(defn pose? [value]
+  (instance? Pose value))
+
+(def default-translation (vector-of :double 0.0 0.0 0.0))
+(def default-rotation (vector-of :double 0.0 0.0 0.0 1.0))
+(def default-scale (vector-of :double 1.0 1.0 1.0))
+(def ^Pose default (->Pose default-translation default-rotation default-scale))
+
+(def ^:private mat4-identity (doto (Matrix4d.) (.setIdentity)))
+(def ^:private num4-zero (vector-of :double 0.0 0.0 0.0 0.0))
+
+(defn- significant-translation? [value]
+  (if-some [[^double x ^double y ^double z] value]
+    (not (and (zero? x)
+              (zero? y)
+              (zero? z)))
+    false))
+
+(defn- significant-rotation? [value]
+  (if-some [[^double x ^double y ^double z ^double w] value]
+    (not (and (zero? x)
+              (zero? y)
+              (zero? z)
+              (== 1.0 w)))
+    false))
+
+(defn- significant-scale? [value]
+  (if-some [[^double x ^double y ^double z] value]
+    (not (and (== 1.0 x)
+              (== 1.0 y)
+              (== 1.0 z)))
+    false))
+
+(defn- add-vector [[^double ax ^double ay ^double az] [^double bx ^double by ^double bz]]
+  (vector-of
+    :double
+    (+ ax bx)
+    (+ ay by)
+    (+ az bz)))
+
+(defn- multiply-vector [[^double ax ^double ay ^double az] [^double bx ^double by ^double bz]]
+  (vector-of
+    :double
+    (* ax bx)
+    (* ay by)
+    (* az bz)))
+
+(defn- rotate-vector [[^double vx ^double vy ^double vz] [^double qx ^double qy ^double qz ^double qw]]
+  (vector-of
+    :double
+    (- (+ (* vx qw qw)
+          (* vy qz qw -2.0)
+          (* vz qy qw 2.0)
+          (* vx qx qx)
+          (* vy qx qy 2.0)
+          (* vz qx qz 2.0))
+       (* vx qy qy)
+       (* vx qz qz))
+    (- (+ (* vx qx qy 2.0)
+          (* vy qy qy)
+          (* vz qy qz 2.0)
+          (* vx qz qw 2.0)
+          (* vy qw qw)
+          (* vz qx qw -2.0))
+       (* vy qx qx)
+       (* vy qz qz))
+    (- (+ (* vx qx qz 2.0)
+          (* vy qy qz 2.0)
+          (* vz qz qz)
+          (* vx qy qw -2.0)
+          (* vy qx qw 2.0)
+          (* vz qw qw))
+       (* vz qx qx)
+       (* vz qy qy))))
```

Comment:
Did you write it yourself? I don't know how to verify this code...

Reply:
I looked at a couple of implementations of the formula, inlined a bunch of stuff, and reordered the coefficients in a way that made sense to me in prefix notation. I double-checked the math against both implementations I looked at to make sure I didn't mess it up, so I'm pretty confident it should be correct, but I guess time will tell! 😅